File: src/main/java/org/genyris/java/swing/GPanel.java
package org.genyris.java.swing;

import java.awt.Graphics;

import javax.swing.JPanel;

import org.genyris.core.Exp;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.AbstractClosure;
import org.genyris.java.JavaUtils;

public class GPanel extends JPanel {

	private static final long serialVersionUID = 1L;
	private AbstractClosure repaint;

	public GPanel(AbstractClosure closure) {
		super();
		repaint = closure;
	}

	public void paintComponent(Graphics g) {
		if (repaint != null) {
			Exp args[] = { JavaUtils.wrapJavaObject(repaint.getEnv(), g) };
			try {
				repaint.applyFunction(repaint.getEnv(), args);
			} catch (GenyrisException e) {
				e.printStackTrace();
			}
		}
	}
}

File: src/org/genyris/java/swing/GPanel.kt
package org.genyris.java.swing

import org.genyris.core.Exp
import org.genyris.exception.GenyrisException
import org.genyris.interp.AbstractClosure
import org.genyris.java.JavaUtils
import java.awt.Graphics
import javax.swing.JPanel

class GPanel(private val repaint: AbstractClosure?) : JPanel() {
    public override fun paintComponent(g: Graphics?) {
        if (repaint != null) {
            val args: Array<Exp?>? = arrayOf<Exp?>(JavaUtils.wrapJavaObject(repaint.getEnv(), g))
            try {
                repaint.applyFunction(repaint.getEnv(), args)
            } catch (e: GenyrisException) {
                e.printStackTrace()
            }
        }
    }

    companion object {
        private const val serialVersionUID = 1L
    }
}
